zokrates_core/src/proof_system/bn128/utils/bellman.rs

```diff
@@ -0,0 +1,494 @@
+                                    .0
+                                    .remove(&var)
+                                    .ok_or(SynthesisError::AssignmentMissing)?
+                                    .into())
+                            },
+                        ),
+                        false => cs.alloc_input(
+                            || format!("PUBLIC_INPUT_{}", index),
+                            || {
+                                Ok(witness
+                                    .0
+                                    .remove(&var)
+                                    .ok_or(SynthesisError::AssignmentMissing)?
+                                    .into())
+                            },
+                        ),
+                    }
+                    .unwrap();
+                    (var.clone(), wire)
+                }),
+        );
+
+        let main = self.main;
+
+        for statement in main.statements {
+            match statement {
+                Statement::Constraint(quad, lin) => {
+                    let a = &bellman_combination(quad.left.clone(), cs, &mut symbols, &mut witness);
+                    let b =
+                        &bellman_combination(quad.right.clone(), cs, &mut symbols, &mut witness);
+                    let c = &bellman_combination(lin, cs, &mut symbols, &mut witness);
+
+                    cs.enforce(|| "Constraint", |lc| lc + a, |lc| lc + b, |lc| lc + c);
+                }
+                _ => {}
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl Computation<FieldPrime> {
+    pub fn prove(self, params: &Parameters<Bn256>) -> Proof<Bn256> {
+        let rng = &mut thread_rng();
+        let proof = create_random_proof(self.clone(), params, rng).unwrap();
+
+        let pvk = prepare_verifying_key(&params.vk);
+
+        // extract public inputs
+        let public_inputs = self.public_inputs_values();
+
+        assert!(verify_proof(&pvk, &proof, &public_inputs).unwrap());
+
+        proof
+    }
+
+    pub fn public_inputs_values(&self) -> Vec<Fr> {
+        self.program
+            .main
+            .arguments
+            .clone()
+            .iter()
+            .zip(self.program.private.clone())
+            .filter(|(_, p)| !p)
+            .map(|(a, _)| a)
+            .map(|v| self.witness.clone().unwrap().0.get(v).unwrap().clone())
+            .chain(self.witness.clone().unwrap().return_values())
+            .map(|v| Fr::from(v.clone()))
+            .collect()
+    }
+
+    pub fn setup(self) -> Parameters<Bn256> {
+        let rng = &mut thread_rng();
+        // run setup phase
+        generate_random_parameters(self, rng).unwrap()
+    }
+}
+
+impl Circuit<Bn256> for Computation<FieldPrime> {
+    fn synthesize<CS: ConstraintSystem<Bn256>>(self, cs: &mut CS) -> Result<(), SynthesisError> {
+        self.program.synthesize(cs, self.witness)
+    }
+}
+
+pub fn serialize_vk(vk: VerifyingKey<Bn256>) -> String {
```

Comment:
This hacky implementation is necessary because the point coordinates are only visible within the pairing crate. Should we consider fixing this in the community edition?